brave/src/main/java/com/linecorp/decaton/processor/runtime/BraveTraceHandle.java

```diff
@@ -0,0 +1,60 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.decaton.processor.runtime;
+
+import com.linecorp.decaton.processor.DecatonProcessor;
+import com.linecorp.decaton.processor.runtime.TracingProvider.TraceHandle;
+
+import brave.Span;
+import brave.messaging.MessagingTracing;
+import brave.propagation.CurrentTraceContext.Scope;
+
+class BraveTraceHandle implements TraceHandle {
+    private final MessagingTracing messagingTracing;
+    private final Span span;
+    private Scope scope;
+
+    public BraveTraceHandle(MessagingTracing messagingTracing, Span span) {
+        this.messagingTracing = messagingTracing;
+        this.span = span;
+    }
+
+    @Override
+    public void processingStart() {
+        scope = messagingTracing.tracing().currentTraceContext().newScope(span.context());
+        span.annotate("decaton.start");
```

Comment:
any reason you renamed this? 
`decaton.start` and `.end` sounds like it covers the whole part of processing inside decaton (so internal queuing etc as well) and confusing I think.

Reply:
The same method is called in the "parent" case (where it represents the start of the pipeline) and the "child" case (where it represents the start of an individual processor), so I thought `decaton.pipeline.start` was misleading. Maybe just `start` and `return` would be clearer?

Reply:
Ah I see. 

I agree with stripping "decaton." prefix because its already scoped by span I think.
However I still think "start" "return" wouldn't be preferred because if I see those annotations in trace the first question would be "what did it started??". So I'd suggest "process.call" "process.return" or something similar that we can tell that it's about process method of processor.

Reply:
> So I'd suggest "process.call" "process.return" or something similar that we can tell that it's about process method of processor.

The problem is that it could be either - in the "parent" span it's `pipeline.return` while in the "child" span it's `process.return`. I think it's ok to have the same method / annotation because the span gives it context, but the annotation name needs to be something that works for both cases.

> Besides, is it still meaningful to have .start annotation? I think it's timing is almost the same as its belonging span's creation so it sounds like duplicating the same information.

I agree that the annotation is not very meaningful for processors, but it's important in the "parent" case, so I think it's easier to keep it consistent.

Reply:
Ah - I see your point.
So what's the purpose of keeping that "parent" code calling `processingStart()` and `processingReturn()`?
`ProcessPipeline#process` calls `context.push(task.taskData());` immediately after calling `processingStart()` so there's no point to keep the call in `ProcessPipeline` I guess?
Then the first (top) processor becomes the direct child of the root span ("decaton") and 2nd processor (if exists) becomes the child of the first processor's span which sounds like exactly what we want.

Reply:
At an internal-API level we need `processingStart` and `processingReturn` around the sync processing section to indicate where to bind the span to the current thread (since the thread that fetches from Kafka is not the thread that runs processing). For the "child" case the thread that created the span is also the thread that runs processing, but I think it's clearest to keep the API consistent.

At a zipkin trace level, the `return` annotation is useful to tell us where sync processing finished (I think both for the overall pipeline and for individual processors). But I guess the `start` annotation isn't so useful anymore: we can tell when deserialisation etc. finished by looking at when the child span started.

I'll remove the `start` annotation and rename `decaton.return` -> `return`?

Reply:
Hmm, I don't understand.

> At an internal-API level we need processingStart and processingReturn around the sync processing section to indicate where to bind the span to the current thread (since the thread that fetches from Kafka is not the thread that runs processing). For the "child" case the thread that created the span is also the thread that runs processing, but I think it's clearest to keep the API consistent.

Even if we delete `processStart` and `processReturn` at `ProcessorPipeline`, the `processStart` call at `ProcessingContextImpl` just works as the point to attach the context to the current thread right?
Because what tracing drivers wants to measure separately from "decaton's internal processing" is the duration of each processor's processing time as you said, when we consider the optimal interface, there should be methods that are called just once before a processor starts processing and just once after a processor returns from processing I think, and I don't understand how having another point of call them at `ProcessorPipeline` could be "consistent" from API's point of view. Rather it feels like reusing a method prepared for another purpose.

> (I think both for the overall pipeline and for individual processors).

How useful it is while two annotations (one from pipeline, one from the top processor) points almost exactly at the same point?

To confirm my understanding, with your current implementation the graphs will be like follows right?
The first `.start` `.return` annotations from pipeline and the another pair from the 1st processor shows almost the same point (there's few instructions between the points but that should be negligible and actually invisible).
```
[--                                                            "decaton" span                                                --]
                                                       ^.start                                                                ^.return // from ProcessorPipeline
<- decaton internal moment->     [--                   top processor's span                         --]
                                                       ^start                                                                 ^return  // from the top processor
                                                                            [--                   2nd processor's span   --]
                                                                            ^start                                            ^return
```

>  But I guess the start annotation isn't so useful anymore: we can tell when deserialisation etc. finished by looking at when the child span started.

Agree with this.

Reply:
> Even if we delete processStart and processReturn at ProcessorPipeline, the processStart call at ProcessingContextImpl just works as the point to attach the context to the current thread right?
> Because what tracing drivers wants to measure separately from "decaton's internal processing" is the duration of each processor's processing time as you said, when we consider the optimal interface, there should be methods that are called just once before a processor starts processing and just once after a processor returns from processing I think, and I don't understand how having another point of call them at ProcessorPipeline could be "consistent" from API's point of view. Rather it feels like reusing a method prepared for another purpose.

Hmm. In terms of how the brave API is intended to be used, spans are meant to be bound to threads as much as possible. But concretely I guess most of the execution of `ProcessPipeline` is covered by the child processors and so binding the parent span is redundant. Fair enough.

In that case I think a full split between parent/child span interfaces makes sense.

> How useful it is while two annotations (one from pipeline, one from the top processor) points almost exactly at the same point?

In the UI we're currently using it's not possible to view annotations from more than one span at the same time. So it's good if each span has whatever annotations are needed to understand that span's behaviour without having to switch back and forth (indeed I'm second-guessing removing the annotation that's at the same place as where the child span starts, since that's not actually so visible when looking at the details of the parent span). But I guess that's a limitation of this specific UI and maybe decaton can't be expected to work around it.

Reply:
>  I guess most of the execution of ProcessPipeline is covered by the child processors

If you mean few lines of code like  by "a few of execution" (in contrast to "most of the execution")) here, it should be totally negligible and pointless to measure as I wrote: 

> (there's few instructions between the points but that should be negligible and actually invisible)

> In the UI we're currently using it's not possible to view annotations from more than one span at the same time. So it's good if each span has whatever annotations are needed to understand that span's behaviour without having to switch back and forth 

Yeah, but the thing is, the ".start" annotation marked on the root span at `ProcessorPipeline` and the starting point of the child span created for the 1st processor points exactly the same time (there can be few microseconds or nanoseconds between, because there are few jvm instructions indeed) and so I think we don't even have to refer the annotation to see the starting point of the processor pipeline. For the `.return` annotation that might not be the case, but the issue sounds trivial enough.

Reply:
> Yeah, but the thing is, the ".start" annotation marked on the root span at ProcessorPipeline and the starting point of the child span created for the 1st processor points exactly the same time (there can be few microseconds or nanoseconds between, because there are few jvm instructions indeed) and so I think we don't even have to refer the annotation to see the starting point of the processor pipeline.

Right, but when looking at the parent span it's actually quite non-obvious where the starting point of the child span is. (And annotations on the child span are not visible at all without clicking back and forth). I agree that it's trivial on a technical level, but it's actually quite a serious difference in usability when a lot of the value of these tools comes from being able to get a visual overview of where execution time is being spent without having to calculate any numbers or crossreference anything.

But I don't think that's decaton's responsibility, and I've followed your approach with the most recent commit - please take a look and see what you think.